pftp/data_handler.go

```diff
@@ -0,0 +1,524 @@
+package pftp
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type dataHandler struct {
+	clientConn   connector
+	originConn   connector
+	config       *config
+	log          *logger
+	waitResponse chan bool
+	connected    chan bool
+}
+
+type connector struct {
+	listener    *net.TCPListener
+	conn        net.Conn
+	remoteIP    string
+	remotePort  string
+	localIP     string
+	localPort   string
+	needsListen bool
+	isClient    bool
+	mode        string
+	connDone    chan error
+}
+
+// Make listener for data connection
+func newDataHandler(config *config, log *logger, clientConn net.Conn, originConn net.Conn, mode string) (*dataHandler, error) {
+	var err error
+
+	d := &dataHandler{
+		originConn: connector{
+			listener:    nil,
+			conn:        nil,
+			needsListen: false,
+			isClient:    false,
+			mode:        config.TransferMode,
+			connDone:    make(chan error),
+		},
+		clientConn: connector{
+			listener:    nil,
+			conn:        nil,
+			needsListen: false,
+			isClient:    true,
+			mode:        mode,
+			connDone:    make(chan error),
+		},
+		config:       config,
+		log:          log,
+		waitResponse: make(chan bool),
+		connected:    make(chan bool),
+	}
+
+	if originConn != nil {
+		d.originConn.remoteIP, _, _ = net.SplitHostPort(originConn.RemoteAddr().String())
+		d.originConn.localIP, d.originConn.localPort, _ = net.SplitHostPort(originConn.LocalAddr().String())
+	}
+
+	if clientConn != nil {
+		d.clientConn.remoteIP, _, _ = net.SplitHostPort(clientConn.RemoteAddr().String())
+		d.clientConn.localIP, d.clientConn.localPort, _ = net.SplitHostPort(clientConn.LocalAddr().String())
+	}
+
+	// When connections are nil, will not set listener
+	if clientConn == nil || originConn == nil {
+		return d, nil
+	}
+
+	if d.clientConn.mode != "PORT" && d.clientConn.mode != "EPRT" {
```

Comment:
what does mean?
could you split this function to method.

Reply:
It just judge pftp needs to open listen port.

Is it better about split different method?

Reply:
Yes.would you split diffrent method please
I think that We should know what you are doing by method name.

Reply:
I had split this branch to another function.

https://github.com/pyama86/pftp/pull/32/commits/64d0a656423f379368945d1390931e04ce254a9d